style-guide.md

```diff
@@ -0,0 +1,65 @@
+- All tasks must have configurable memory and disk space allocations
+  - TODO: only enforceable if we require specific var naming. Do we want to enforce specific variable names? Keep this in the guide?
+  - see the various tasks in the template directory for possible ways to allocate resources
+    - Contributors can mix and match the available templates, copy and pasting subsections as appropriate
+    - It is allowed to have one resource allocated dynamically, and another allocated statically in the same task.
+    - It is *not* allowed to have a resource which can be allocated *either* statically or dynamically.
+      - This is technically feasible, but is too complicated for maintenance and end-users.
+      - e.g. `memory_gb` and `modify_memory_gb` cannot be present in the same task.
+- All tasks and workflows should have a `max_retries` input.
+  - TODO: again, only enforeceable with specific var names. Not sure if we want to get into that business. Keep it in the guide?
+  - This should be defaulted to `1` for nearly all tasks
+  - Some tasks are particularly error prone and can have a higher default `max_retries`
+  - **rule specific to workflows:** `max_retries` should be an optional `Int?`
+    - This allows each task to have it's own specific default `max_retries`
+      - If a user does not supply `max_retries`, those task level defaults will get used
+      - If a user does supply `max_retries`, it should override the default for *every* task called
+- multi-core tasks should *always* follow the conventions laid out in the `detect_nproc_task` example (see `template/task-templates.wdl`)
+  - this is catering to cloud users, who may be allocated a machine with more cores than are specified by the `ncpu` parameter
+  - TODO: this would be tough to validate with a linter. It is definitely a best practice, but IMO doesn't belong in the guide
+- `command` blocks should be wrapped with arrows (`<<< >>>`) instead of brackets (`{ }`)
+  - Certain Bash constructions cause problems with the bracket notation
+- output file names should *always* be determined with either the `outfile_name` parameter or the `prefix` parameter.
+  - `outfile_name` should be preferred if no downstream tasks/tools rely on the file name/extension
+  - tasks with multiple outputs should always use the `prefix` convention
+  - TODO: IMO another "best practice" that shouldn't be enforced by a linter. Remove from guide?
+- output variable names should be short but descriptive
+  - TODO: This I'm more inclined to enforce with the linter. I think disallowing specific output names is better than forcing users into using specific ones.
+  - Disallowed output names: `out*`, any name less than 3 characters
+    - TODO: is this sufficient?
+- All tasks should run in a Docker container
+  - TODO: is there a valid reason not to use a Docker container? Should this be enforced by the linter?
```

Comment:
me and @adthrasher discussed this offline. We've decided there could be a use case where a user is unconcerned with a "reproducible" workflow, and just wants it to work on their customized cluster environment.

So the approach we decided to take is to leave this line in the style guide, and by default enforce that every task has a docker image with the linter, but provide a CL flag for disabling that specific check. It'll be up to the user if they want to ignore this rule, but we should encourage them not to.